### PR TITLE
Time - Hannah J.

### PR DESCRIPTION
**Pull request by Hannah0124**
# Assignment Submission: Hotel

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? | I struggled to determine how many classes I should make. Also, it was hard to decide whether or not to make object compositions and class inheritance. 
What was a design decision you made that changed over time over the project? | I kept changing my decision over whether or not store `reservations` in the `HotelManager` class as an Array or a Hash. In particular, I had difficulty deciding a relationship between `Reservation` and `Block` class. 
What was a concept you gained clarity on, or a learning that you'd like to share? | I learned that I can invoke an instance method in the constructor.
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? | I added a nominal test case to check the correct `#total_cost` in the `reservation.test.rb`.
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? | When there is no reservation result, I made sure to raise an ArgumentError for the `#find_reservations` method.
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | Personally, I think that writing pseudocode is tough. It's definitely an area in which I need to keep pushing myself, but drawing diagrams helped me a lot to think of the logic and the design decision. Futhermore, it was tough for me to write down from tests to codes since it takes more time to think about the logic in advance. However. I did my best to keep the orders. By the time I had to change all the written codes, I was frustrated considering that I had to re-write all the test files from the beginning. Plus, not having knowledge of how to use a debugging tool made it more time-consuming to make my program.

### CODE REVIEW

File: lib/date_range.rb [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  class DateRange 
+
+    attr_reader :start_date, :end_date 
+
+    def initialize(start_date, end_date)
+      @start_date = start_date
+      @end_date = end_date
+
+      # Invoke the method
+      check_valid_date
```
Comment (beccaelenzil) [inline lib/date_range.rb:14]: Good use a custom helper method

File: lib/date_range.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    end 
+
+    
+    def nights 
+      return @end_date.day - @start_date.day
+    end 
+
+    # When searching for the specific date
+    # if the new date is between start and end date (excluding end_date)
+    def include?(date)
```
Comment (beccaelenzil) [inline lib/date_range.rb]: Nice work using a boolean expression. Remember to use explicit return

File: lib/hotel_manager.rb [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+      
+      self.add_reservation(reservation, type: :individual)
+
+      return reservation
+    end 
+
+
+    #  Find reservations per room withinn specific date range (using overlap?)
+    def reservations_per_room(date_range, room)
+      return @reservations[:individual].filter do |reservation|
```
Comment (beccaelenzil) [inline lib/hotel_manager.rb]: Nice use of an enumerable method.

File: lib/hotel_manager.rb [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+
+    def available_blocks(date_range, number_of_rooms:)
+      return @blocks.find_all do |block|
+        block.date_range.same?(date_range) && (block.rooms.length == number_of_rooms)
+      end 
+    end 
+
+    
+    def make_block_reservation(date_range, number_of_rooms: 1)
```
Comment (beccaelenzil) [inline lib/hotel_manager.rb]: Per user stories, blocks need to have a name or special id associated with them. I'm not sure I see how this functionality is built in. I think this would just be a matter of adding an attribute to the block class. Or maybe this is accounted for with the reservation id? "These rooms are set aside, and are made available for reservation by certain customers at a discounted rate."

File: test/hotel_manager_test.rb [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+        expect(reservations).must_be_instance_of Array
+        expect(reservations[0]).must_be_instance_of Hotel::Reservation
+        expect(reservations[0].room.id).must_equal 8
+      end 
+    end 
+  end
+
+  describe "wave 2" do 
+    describe "#available_room_ids" do 
+      it "returns available rooms for the given date range" do
```
Comment (beccaelenzil) [inline test/hotel_manager_test.rb]: Make sure to also test for the edge case where there are no available room.

File: test/hotel_manager_test.rb [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+
+      it "raises an ArgumentError if the number of rooms is not between 1 and 5" do 
+
+        expect{@hotel_manager.make_block_reservation(@date_range, number_of_rooms: 6)}.must_raise ArgumentError
+
+        expect{@hotel_manager.make_block_reservation(@date_range, number_of_rooms: 0)}.must_raise ArgumentError
+      end 
+
+
+      it "raise an ArgumentError if there are no available rooms for block" do
```
Comment (beccaelenzil) [inline test/hotel_manager_test.rb]: It is also important to test that it raises in exception if there are no available rooms because the rooms are taken up by normal reservations. I believe you have implemented this functionality correctly, but it's hard to know for sure without a unit test.

**Comment by beccaelenzil** [comment]
# Hotel






## Section 1: Major Learning Goals



| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
Practices SRP by having at least two separate classes with distinct responsibilities, and test files for these two classes | ✔️
Overall, demonstrates understanding instance variables vs. local variables. (There aren't unnecessarily too many instance variables, when it should be a local variable) | ✔️
For each test file, tests demonstrate an understanding of instantiating objects properly, and using Arrange-Act-Assert | ✔️
Practices pseudocode and TDD, and reflected on it by filling out the reflection questions | ✔️
Practices git with at least 15 small commits and meaningful commit messages | ✔️

## Section 2: Code Review and Testing Requirements

| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
There is a class that represents a reservation, and a second class that holds/manages a collection of reservations through composition (instance variable) | ✔️
The logic for checking if a reservation's date overlaps with another reservation's date is complex logic that is separated into method(s) (and potentially class(es)) | ✔️
The logic for checking if a reservation's date overlaps with another reservation's date has unit tests | ✔️
All of the given tests run and pass | ✔️
A test coverage tool is installed and used, and shows 95% test coverage | ✔️

## Section 3: Feature Requirements

| Feature Requirement: There is a method that... | yes/no |
| --- | --- |
gives back a list of rooms, and it's tested | ✔️
creates a specific reservation for a room for a given date range, and it has nominal test cases | ✔️
creates a specific reservation for a room for a given date range, and it tests an edge case, such as no available room, or invalid date range | ✔️
gives back a list of reservations on a given date. Its tests include a test that makes several reservations for a given date | ✔️
calculates the total price for a reservation | ✔️
gives back a list of available rooms for a given date range, and it has nominal test cases | ✔️
gives back a list of available rooms for a given date range, and it has edge test cases, such as no available room, or invalid date range | ✔️
creates a block of rooms | ✔️
reserves a room from a block | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ total in all sections |


### Additional Feedback



Great work overall! You've built your first project with minimal starting code. This represents an incredible milestone in your journey, and you should be proud of yourself!

I am particularly impressed by the way that you that you made design decision to work through the tricky logic of implementing blocks. In addition, your tests are quite comprehensive. I've made a few suggestions about edge case tests. It is clear that the learning goals around object oriented design and TDD were met. Keep up the hard work!

## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Logical/Organized | ✅

File: lib/hotel_manager.rb (hunk 4 above)
Comment (Hannah0124) [reply lib/hotel_manager.rb]: Thank you Becca for checking them in detail. 🙏 That was one of my challenges and I felt that my design might be too complicated. Like you guessed, for the latter part ("maybe this is accounted for with the reservation id? "), I made a Block connected with Reservation. For instance, (1) if it's a block, Reservation has a block and the Block has rooms, whereas (2) If it's a regular reservation, Reservation has a room instead of Block. 

<Summary>

**AS IS**

- Block: Reservation -> Block -> Rooms
- Individual: Reservation -> Room

My question is "Would it be better to separate BlockReservation and Reservation and make them like Polymorphism as written below"? 

**TO BE**

- BlockReservation -> Rooms 
- Reservation -> Rooms

I really appreciate your feedback! Thank you. 🥰
